test: automatically install `anvil` during `yarn install` via `foundryup.ts`

davidmurdoch opened the pull request:
I've ported Foundry's foundryup bash script to TypeScript. It downloads a pinned version of the `anvil` binary on `yarn install`.

While you don't need to do anything, other than run `yarn` (or `yarn install`), you can customize the installation options by running `yarn foundryup <options>`. Try running `yarn foundryup --help` to see all that it can do.

An explanation of how it works:

The `yarn foundryup` command caches (to `.metamask/cache/*`) the downloaded binaries (defaults to only `anvil`, but it _can_ install all the others) and symlinks them from the cache into `node_modules/.bin/<binary name>`. This makes successive installs nearly instant.

In CI the cache itself is cached (same process as for the `.yarn/cache` cache), so that CI doesn't have to redownload binaries on every workflow.

The `foundryup` code is intended to be generic, and could eventually be a partial replacement for the `ganache` package for most users (for cli users only, not for programmatic use cases). For this reason, the script uses defaults that are not tailored to our needs; instead we use a new `package.json` property _`foundryup`_ to describe what defaults we do want, and the script uses those values.

In addition to the original `foundryup` bash script features, this script supports checksumming the downloaded binaries; a feature requested by the MM security team. See the `foundryup` property in our `package.json` for more details.

A feature that is _missing_ from this script, but is in the original `foundryup` bash script,is support for downloading the `man` pages. I don't really care about this feature and don't think anyone would use it.

This script _should_ support the following platforms, but I've only tested Linux+AMD:

Linux on ARM and AMD
Mac on ARM and AMD
Windows on AMD (likely works on arm-based Windows machines via OS's emulation)

--

This also adds a new yarn command: `anvil`. You can use it just like you did with `yarn ganache`, i.e., `yarn anvil`; note: the defaults and CLI options are not going to be exactly the same as `ganache`.


<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.


## **Description**


[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28393?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to this page...
2.
3.

## **Screenshots/Recordings**

### **Before**


### **After**


## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

-->

davidmurdoch committed:
test: automatically install `anvil` during `yarn install` via `foundryup.ts`

davidmurdoch commented on package.json:
```diff
@@ -687,7 +692,7 @@
   },
   "engines": {
-    "node": ">= 20",
+    "node": ">= 20.17",
```
`foundryup.mts` uses node feature introduced in `20.17`

davidmurdoch commented on .circleci/config.yml:
```diff
@@ -3,25 +3,25 @@ version: 2.1
       NODE_OPTIONS: --max_old_space_size=2048
   node-browsers-medium:
     docker:
-      - image: cimg/node:20.17-browsers
```
20.17 doesn't match `ubuntu-2404:2024.05.1` (aka `ubuntu-2404:current`), so i've updated to so the versions match.

davidmurdoch commented on .nvmrc:
```diff
@@ -1 +1 @@
-v20.17
+v20.12.0
```
update to match CI

davidmurdoch commented on package.json:
```diff
@@ -687,7 +692,7 @@
   },
   "engines": {
-    "node": ">= 20",
+    "node": ">20.12.0 <20.15.0 || >=20.17.0",
```
```suggestion
    "node": ">=20.12.0 <20.15.0 || >=20.17.0",
```

davidmurdoch commented on .circleci/config.yml:
```diff
@@ -449,12 +449,15 @@ jobs:
           command: yarn --version > /tmp/YARN_VERSION
+      - run:
+          name: Save Foundry version
+          command: node -e "process.stdout.write(require('./package.json').foundry)" > /tmp/FOUNDRY_VERSION
```
So I've added a new field to our `package.json`: `foundry`. It contains the release we use in the installation process, as well as in this cache process.

I don't like it.

Please please please recommend a better way <3

davidmurdoch commented on .circleci/config.yml:
```diff
@@ -466,12 +469,14 @@ jobs:
+          key: dependency-cache-{{ checksum "/tmp/YARN_VERSION" }}-{{ checksum "yarn.lock" }}-{{ checksum "/tmp/FOUNDRY_VERSION" }}
           paths:
             - .yarn/cache
+            - .metamask/cache # should match yarn's relative location
```
Adds the anvil binary to the CI cache.

davidmurdoch commented on .circleci/config.yml:
```diff
@@ -466,12 +469,14 @@ jobs:
       - persist_to_workspace:
           root: .
           paths:
+            - .metamask/cache # ensures anvil is installed
```
Now that I'm doing a self review, I'm thinking this probably isn't required, since the `anvil` binary we need is symlinked to `.metamask/cache` from within in `node_modules/.bin`, and I think `persist_to_workspace` will copy the symlinks automatically when it copies `node_modules`.

Can't hurt to keep it though?

davidmurdoch commented on .gitignore:
```diff
@@ -64,6 +64,9 @@ test-results/
 development/generate-attributions/.yarn/*
 
+# MetaMask
+.metamask/*
```
I cache the foundry foundries in a new `.metamask` root folder. we want to ignore it.

Maybe this is where I could put the foundry version? instead of the package.json? hmmmm. not sure I like that either

seaona replied:
mmm I think I would rather keep the version in the package.json... just because might be more intuitive? :thinking: not sure though

davidmurdoch commented on package.json:
```diff
@@ -755,5 +760,6 @@
   },
-  "packageManager": "yarn@4.5.1"
+  "packageManager": "yarn@4.5.1",
+  "foundry": "nightly-31dd1f77fd9156d09836486d97963cec7f555343"
```
So groossssss

seaona replied:
:cry: there's not much we can do about it... it doesn't seem they have semantic versioning? :thinking:

davidmurdoch commented on package.json:
```diff
@@ -9,7 +9,8 @@
     "webpack:clearcache": "./development/clear-webpack-cache.js",
-    "postinstall": "yarn webpack:clearcache",
+    "foundryup": "tsx --tsconfig ./test/helpers/foundry/tsconfig.json ./test/helpers/foundry/foundryup.mts",
+    "postinstall": "yarn webpack:clearcache && yarn foundryup --binaries anvil --version \"$(node -e \"process.stdout.write(require('./package.json').foundry)\")\"",
```
This is another place I use the `package.json`'s `foundry` property (the other was in the circleci yaml). Have you thought of a better way to do this yet?

davidmurdoch replied:
Should we hardcode checksums of the binaries we expect somewhere in our repo, and then verify that the binaries are correct. `foundryup` (both this script or the original bash script), doesn't have the version immutability guarantee that `npm`/`yarn` do 🤔, so it makes our developers (and thus users, by a stretch) susceptible to supply chain attacks. 


cc @kumavis @naugtur: a summary: this PR adds foundry's `anvil` to the extension; it is a binary downloaded from Foundry's Github Releases page, so it can be swapped out at any time.

naugtur replied:
How important is keping it up to date? Could we install it once and keep the binary in the repo? It's easier and less likely to be ignored. If we do checksums, people will just update checksums and be annoyed ;)

davidmurdoch replied:
> How important is keping it up to date?

This is to replace Ganache, a package that hasn't been updated in over a year now. So apparently we don't need it to be updated all too often. :-)

davidmurdoch commented on test/helpers/foundry/foundryup.mts:
```diff
@@ -0,0 +1,95 @@
+
+const parsedArgs = parseArgs();
+
+const { enableGlobalCache } = parseYaml(readFileSync('.yarnrc.yml', 'utf-8'));
+const CACHE_DIR = enableGlobalCache
+  ? join(homedir(), '.cache', 'metamask')
+  : join(cwd(), '.metamask', 'cache');
```
Probably not necessary, but this ensures that we keep our own cache alongside the `.yarn` cache. Why? I had to put it somewhere, and copying whatever we do for yarn just felt right, I guess.

davidmurdoch commented on test/helpers/foundry/foundryup.mts:
```diff
@@ -0,0 +1,95 @@
+  : join(cwd(), '.metamask', 'cache');
+
+if (parsedArgs.command === 'cache clean') {
+  await rm(CACHE_DIR, { recursive: true, force: true });
```
Have a `cache clean` was mostly just useful when writing this code and manually testing it. Feature could be removed if we don't want to keep code we hopefully won't need.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+      coerce: (
+        rawVersion: string,
+      ): { version: 'nightly' | `v${string}`; tag: string } => {
+        if (/^nightly/u.test(rawVersion)) {
+          return { version: 'nightly', tag: rawVersion };
+          // we don't validate the version much, we just trust the user
+        } else if (/^\d/u.test(rawVersion)) {
+          return { version: `v${rawVersion}`, tag: rawVersion };
+        }
```
This chunk was ported from the bash script. It's weird though, because foundry doesn't publish tagged versions... only `nightly-<commit hash>` and `nightly` tags.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+Support    : https://t.me/foundry_support/
+Contribute : https://github.com/orgs/foundry-rs/projects/2/
+
+.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx
```
This is a bit obnoxious. I wouldn't hate it if you asked me to remove it. :-)

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+export async function extractFrom(url: URL, binaries: string[], dir: string) {
+  const extract = url.pathname.toLowerCase().endsWith(BinFormat.Tar)
+    ? extractFromTar
+    : extractFromZip;
```
Windows binaries are packaged in a `.zip`, mac and linux use `.tar.gz`

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+ */
+async function extractFromTar(url: URL, binaries: string[], dir: string) {
+  const downloads: string[] = [];
+  await pipeline(
```
Promise pipelines are really cool.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+  const downloads: string[] = [];
+  await pipeline(
+    startDownload(url),
+    extractTar({ cwd: dir }, binaries).on('entry', ({ absolute }) =>
```
We can use `extractTar` to only extract the binaries we asked for. So we do. Zoom zoom.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+  });
+  const source: Source = {
+    async size() {
+      const download = startDownload(url, { agent, method: 'HEAD' });
```
`download` is a `PassThrough` stream, which, for us, is just a stream that acts a middleman between the source stream and a destination stream.

I'll explain a bit more in the `startDownload` function below.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+      const options = {
+        agent,
+        headers: {
+          range: `bytes=${offset}-${bytes ? offset + bytes : ''}`,
```
Grab only certain parts of the file from github; no need to download the whole thing!

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+  };
+
+  const { files } = await Unzip.custom(source);
+  const filtered = files.filter(({ path }) =>
+    binaries.includes(basename(path, extname(path))),
+  );
```
Gets the _index_ to the binaries we care about (no files have been downloaded yet)

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+    binaries.includes(basename(path, extname(path))),
+  );
+  return await Promise.all(
+    filtered.map(async ({ stream, path }) => {
+      const dest = join(dir, path);
+      await pipeline(stream(), createWriteStream(dest));
+      return dest;
+    }),
```
Now we can download only the binaries we care about.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+      stream.emit('response', response);
+
+      response.once('error', stream.emit.bind(stream, 'error'));
+      await pipeline(response, stream).catch(stream.emit.bind(stream, 'error'));
```
More cool `pipeline` usage!

the call to `startDownload` created the `stream` earlier (a `PassThrough` steam), and returned it immediately even thought we hadn't actually downloaded anything yet. Now that we have started receiving data we just send it along.

Notice that we also emit errors on the `steam` object when appropriate.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+}
+
+export class DownloadStream extends Stream.PassThrough {
+  async response(): Promise<IncomingMessage> {
+    return new Promise((resolve, reject) => {
+      this.once('response', resolve);
+      this.once('error', reject);
+    });
+  }
```
In a `HEAD` only request, there is no `body`, so there is no data to send to the stream. This `response` function is used to return the headers (and status, statusCode, etc).

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+ * @returns The list of binaries extracted
+ */
+async function extractFromZip(url: URL, binaries: string[], dir: string) {
+  const agent = new (url.protocol === 'http:' ? HttpAgent : HttpsAgent)({
+    keepAlive: true,
+  });
```
We end up querying the same connection multiple times, so keeping the connection alive in between queries improves performance.

davidmurdoch commented on types/unzipper.d.ts:
```diff
@@ -0,0 +1,17 @@
+import 'unzipper';
+
+declare module 'unzipper' {
```
added some missing types to the `unzipper` package

seaona commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+      alias: 'v',
+      description:
+        'Specify the version (see: https://github.com/foundry-rs/foundry/tags)',
+      default: 'nightly',
```
should we default to nightly, or should we set a fixed version and update it in separate PRs periodically, to make sure nothing is broken with a version update?

davidmurdoch replied:
The install script I added in `package.json` uses a pinned version. This script defaults to nightly; eventually I think we'll to release this code to npm as a stand-alone package for other developers to use in the future, so I think `nightly` would make sense when we do that, so we should keep it as nightly here. Just my opinion.

seaona replied:
oh I see, it make sense, thank you for clarifying :pray:

seaona commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+
+export function printBanner() {
+  console.log(`
+.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx
+
+ ╔═╗ ╔═╗ ╦ ╦ ╔╗╔ ╔╦╗ ╦═╗ ╦ ╦         Portable and modular toolkit
+ ╠╣  ║ ║ ║ ║ ║║║  ║║ ╠╦╝ ╚╦╝    for Ethereum Application Development
+ ╚   ╚═╝ ╚═╝ ╝╚╝ ═╩╝ ╩╚═  ╩                 written in Rust.
+
+.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx
+
+Repo       : https://github.com/foundry-rs/
+Book       : https://book.getfoundry.sh/
+Chat       : https://t.me/foundry_rs/
+Support    : https://t.me/foundry_support/
+Contribute : https://github.com/orgs/foundry-rs/projects/2/
+
+.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx.xOx
```
hahahah I'm fine with it 🤭

seaona commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,393 @@
+export function parseArgs(args: string[] = argv.slice(2)) {
+  const { $0, _, ...parsed } = yargs()
+    // Ensure unrecognized commands/options are reported as errors.
+    .strict()
```
who didn't know about this!

seaona commented on test/helpers/foundry/foundryup.mts:
```diff
@@ -0,0 +1,95 @@
+} from './helpers.mts';
+
+const parsedArgs = parseArgs();
+
```
maybe we could split the different logic into smaller functions? From chatgpt could be something like:

- getCacheDirectory
- getBinaryArchiveUrl
- cleanCache
- checkAndDownloadBinaries
- installBinaries
- main

davidmurdoch replied:
Sure, most of those make sense.

I do think over-abstraction is a real concern, and can lead to reduce readability in an effort to improve it. Like `cleanCache`, for example, is already a short bit of code that is straight forward to read, moving the logic to its own file and function would increase the lines of code from 1 to about 5 (including the import and calling the function, and maybe more if adding JSDoc comments).

Feel free to add functions where you see fit, just don't blindly listen to all of ChatGPTs recommendations, I think it goes a bit too far sometimes.

seaona replied:
I see, your point make sense to me! I've added a proposal for split in this PR, which also add unit tests. Let me know if this makes sense, or if you prefer to have it encapsulated with less functions :pray: 

https://github.com/MetaMask/metamask-extension/pull/28833

davidmurdoch committed:
add checksumming to foundryup binaries

davidmurdoch committed:
update CI

davidmurdoch committed:
depcheck fix

davidmurdoch committed:
polish

davidmurdoch commented on package.json:
```diff
@@ -660,13 +664,15 @@
     "through2": "^4.0.2",
     "ts-node": "^10.9.2",
-    "tsx": "^4.7.1",
+    "tsx": "^4.19.2",
```
This has been updated to fix stack traces on Node 20+

davidmurdoch commented on package.json:
```diff
@@ -754,5 +760,23 @@
+      "algorithm": "sha256",
+      "binaries": {
+        "anvil": {
+          "darwin-amd64": "6ad1c31bb8bc8c78551792ad538028385eb5a44d105f9af40ddeb1b75c1327b1",
+          "darwin-arm64": "cbb36860b2e08a7d5a35bec430904a1dc0319602ac0923017b81d9a8ffdcc5b0",
+          "linux-amd64": "56054e7d59547beb6adacbb9060a58595a8c5cca185f946317334ece509e0433",
+          "linux-arm64": "16494ae631d8ef73944abaaaa756661c1d43bbfc7642e7fff0a0d807d6d143c4",
+          "win32-amd64": "2034bdf35023e59c5dc06d6717a8eb2cede1204dbc60fd2507564422f03dceb3"
```
Yes, these are slightly annoying to have to generate :-)

seaona replied:
mmm maybe we can add a script in package json to help generate this for future versions? :thinking: 
non-blocking/could be done later on

davidmurdoch commented on package.json:
```diff
@@ -754,5 +760,23 @@
+        }
+      }
+    },
+    "version": "nightly-31dd1f77fd9156d09836486d97963cec7f555343"
```
foundry doesn't publish versions, just tags like this. it's weird, but it's all we got.

naugtur replied:
frankly, commit hash is better for us than a tag

davidmurdoch replied:
@naugtur commit hash doesn't work because we don't download the assets from `git`. The assets are only stored in the _GitHub Release_ page.

We can look up the tags related to a commit (and visa versa), but since releases and tags are mutable that won't help.

HowardBraham replied:
I think this changed after you wrote that On Dec 3.
Here's `0.3.0`: https://github.com/foundry-rs/foundry/releases/tag/v0.3.0
And there's also a `stable`: https://github.com/foundry-rs/foundry/releases/tag/stable

davidmurdoch replied:
Updated to latest stable (v0.3.0)

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,522 @@
+            ? JSON.parse(rawChecksums)
+            : rawChecksums;
+        } catch {
+          throw new Error('Invalid checksums');
```
we could do a lot more validation of the checksum object... but I think it's bit premature, since maybe the way I implemented it is horirble.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,522 @@
+        transform: (entry) => {
+          if (checksumAlgorithm) {
+            const hash = createHash(checksumAlgorithm);
+            const passThrough = new Minipass({ async: true });
```
`Minipass` is almost like Node's `Stream.PassThrough`, but different enough that if we use `PassThrough` the data becomes corrupted (the hash is correct, but the binary is out of order -- I couldn't figure out why it happens). So it is lame that we need a whole new lib for this, but it is what it is.

davidmurdoch commented on test/helpers/foundry/helpers.mts:
```diff
@@ -0,0 +1,522 @@
+      const destStream = createWriteStream(dest);
+      if (checksumAlgorithm) {
+        const hash = createHash(checksumAlgorithm);
+        const hashStream = async function* (source: Entry) {
+          for await (const chunk of source) {
+            hash.update(chunk);
+            yield chunk;
+          }
+        };
```
Instead of `tee`ing the stream I'm just reading the data as it passes through the `pipeline` and then letting it continue on untouched. This could be done with a `Stream.PassThrough`, but that would requires a creating a new `class` that extends `Stream.PassThrough`.

seaona committed:
test: foundryup unit (#28833)

<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been
completely filled out, and PR status checks have passed at least once.
-->

## **Description**
- Splits the logic of foundryup.ts file into functions, so they are
easier to test
- Adds unit tests
- Updates file extensions from `mts` to `ts`. The reason for that is
because otherwise the unit tests don't run out of the box and it was
suggested to change the config file and add a new dependency to make it
work, which seems unnecessary for the scoped task of installing foundry.
Also, this way we get lint out of the box too (instead of having to
update the lint rules to include mts files)

The unit tests can be run with `yarn test:unit
test/helpers/foundry/foundryup.test.ts`

<!--
Write a short description of the changes included in this pull request,
also include relevant motivation and context. Have in mind the following
questions:
1. What is the reason for the change?
2. What is the improvement/solution?
-->

[![Open in GitHub
Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/28833?quickstart=1)

## **Related issues**

Fixes:

## **Manual testing steps**
- Run unit tests: `yarn test:unit
test/helpers/foundry/foundryup.test.ts`
- Install default (anvil with checksum verification) `yarn foundryup`
- Install forge and cast - it should skip the checksum verification
`yarn foundryup -b forge -b cast`
- See commands `yarn foundryup --help`
- Clean cache before installing `yarn foundryup cache clean`
- Pass alternative repo `yarn foundryup --repo alternative-fork/foundry`



## **Screenshots/Recordings**


![Screenshot from 2024-12-10
11-47-37](https://github.com/user-attachments/assets/88893978-aa78-4bf1-b2bb-e17ab3d6748e)


## **Pre-merge author checklist**

- [ ] I've followed [MetaMask Contributor
Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask
Extension Coding
Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format
if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling
guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the
app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described
in the ticket it closes and includes the necessary testing evidence such
as recordings and or screenshots.

---------

Co-authored-by: David Murdoch <187813+davidmurdoch@users.noreply.github.com>

seaona committed:
Merge branch 'main' into foundryup

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+      const destStream = createWriteStream(dest);
+      if (checksumAlgorithm) {
+        const hash = createHash(checksumAlgorithm);
+        const hashStream = async function* (entryStream: Entry) {
+          for await (const chunk of entryStream) {
+            hash.update(chunk);
+            yield chunk;
+          }
+        };
```
Instead of `tee`ing the stream I'm just reading the data as it passes through the `pipeline` and then letting it continue on untouched. This could be done with a `Stream.PassThrough`, but that would requires a creating a new `class` that extends `Stream.PassThrough`.

naugtur replied:
A noop transform would do it

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+            throw new Error('Missing absolute path for entry');
+          }
+
+          if (checksumAlgorithm) {
+            const hash = createHash(checksumAlgorithm);
+            const passThrough = new Minipass({ async: true });
```
`Minipass` is almost like Node's `Stream.PassThrough`, but different enough that if we use `PassThrough` the data becomes corrupted (the hash is correct, but the binary is out of order -- I couldn't figure out why it happens). So it is lame that we need a whole new lib for this, but it is what it is.

naugtur replied:
I might give it a try to rewrite to built-in streams

davidmurdoch commented on test/helpers/foundry/options.ts:
```diff
@@ -0,0 +1,142 @@
+            ? JSON.parse(rawChecksums)
+            : rawChecksums;
+        } catch {
+          throw new Error('Invalid checksums');
```
we could do a lot more validation of the checksum object... but I think it's bit premature, since maybe the way I implemented it is horrible.

davidmurdoch commented on test/helpers/foundry/options.ts:
```diff
@@ -0,0 +1,142 @@
+      coerce: (
+        rawVersion: string,
+      ): { version: 'nightly' | `v${string}`; tag: string } => {
+        if (/^nightly/u.test(rawVersion)) {
+          return { version: 'nightly', tag: rawVersion };
+          // we don't validate the version much, we just trust the user
+        } else if (/^\d/u.test(rawVersion)) {
+          return { version: `v${rawVersion}`, tag: rawVersion };
+        }
```
This chunk was ported from the bash script. It's weird though, because foundry doesn't publish tagged versions... only `nightly-<commit hash>` and `nightly` tags. Kept it here just to maintain feature parity.

davidmurdoch commented on test/helpers/foundry/options.ts:
```diff
@@ -0,0 +1,142 @@
+export function parseArgs(args: string[] = argv.slice(2)) {
+  const { $0, _, ...parsed } = yargs()
+    // Ensure unrecognized commands/options are reported as errors.
+    .strict()
+    // disable yargs's version, as it doesn't make sense here
+    .version(false)
+    // use the scriptName in `--help` output
+    .scriptName('yarn foundryup')
+    // wrap output at a maximum of 120 characters or `stdout.columns`
+    .wrap(Math.min(120, stdout.columns))
+    .parserConfiguration({
+      'strip-aliased': true,
+      'strip-dashed': true,
+    })
+    // enable ENV parsing, which allows the user to specify foundryup options
+    // via environment variables prefixed with `FOUNDRYUP_`
+    .env('FOUNDRYUP')
```
This was just copied and tweaked from the `webpack` yargs setup script.

davidmurdoch commented on test/helpers/foundry/types.ts:
```diff
@@ -0,0 +1,100 @@
+
+export enum Binary {
+  Anvil = 'anvil',
+  Forge = 'forge',
+  Cast = 'cast',
+  Chisel = 'chisel',
```
we could use all these just by adding to `--binaries <binary you want>` to the `yarn foundryup` command. neat!

davidmurdoch commented on test/helpers/foundry/foundryup.ts:
```diff
@@ -0,0 +1,166 @@
+  const CACHE_DIR = getCacheDirectory();
+
+  if (parsedArgs.command === 'cache clean') {
+    await rm(CACHE_DIR, { recursive: true, force: true });
```
Having a `cache clean` was mostly just useful when writing this code and manually testing it. Feature could be removed if we don't want to keep code we hopefully won't need.

davidmurdoch commented on test/helpers/foundry/foundryup.ts:
```diff
@@ -0,0 +1,166 @@
+
+  if (parsedArgs.command === 'cache clean') {
+    await rm(CACHE_DIR, { recursive: true, force: true });
+    say('done!');
```
The `say` function name is copied from the original foundryup bash script. I liked it, so I kept it.

davidmurdoch commented on test/helpers/foundry/foundryup.ts:
```diff
@@ -0,0 +1,166 @@
+import { Architecture, Checksums, Extension, Platform } from './types';
+
+export function getCacheDirectory(): string {
+  let enableGlobalCache = false;
+  try {
+    const configFileContent = readFileSync('.yarnrc.yml', 'utf8');
+    const parsedConfig = parseYaml(configFileContent);
+    enableGlobalCache = parsedConfig?.enableGlobalCache ?? false;
+  } catch (error) {
+    console.error('Error reading/parsing .yarnrc.yml:', error);
+  }
+  return enableGlobalCache
+    ? join(homedir(), '.cache', 'metamask')
+    : join(cwd(), '.metamask', 'cache');
```
Probably not necessary, but this ensures that we keep our own cache alongside the `.yarn` cache. Why? I had to put it somewhere, and copying whatever we do for yarn just felt right, I guess.

davidmurdoch commented on .circleci/config.yml:
```diff
@@ -336,12 +336,15 @@ jobs:
       - run:
           name: Save Yarn version
           command: yarn --version > /tmp/YARN_VERSION
+      - run:
+          name: Save Foundry version
+          command: node -e "process.stdout.write(require('./package.json').foundryup.version)" > /tmp/FOUNDRY_VERSION
```
So I've added a new field to our `package.json`: `foundryup`. It contains the release we use in the installation process, as well as in this cache process.

I don't like it.

Please please please recommend a better way <3

davidmurdoch replied:
For caching, I just copied the way we save the yarn version (by writing to a tmp file). I don't care either way though.

What I like about using `node` where feasible, instead of *nix tools (`fromJson` in this case), is that we know that all developers at MetaMask have node installed, but they might not have `fromJson`. It makes running the commands easier.

I don't care all that much though. Just my own preference.

davidmurdoch commented on package.json:
```diff
@@ -768,5 +774,23 @@
   },
-  "packageManager": "yarn@4.5.1"
+  "packageManager": "yarn@4.5.1",
+  "foundryup": {
```
So groossssss

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+  dir: string,
+  checksums: { algorithm: string; binaries: Record<Binary, string> } | null,
+) {
+  const extract = url.pathname.toLowerCase().endsWith(Extension.Tar)
+    ? extractFromTar
+    : extractFromZip;
```
Windows binaries are packaged in a `.zip`, mac and linux use `.tar.gz`

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+    binary: string;
+    checksum?: string;
+  }[] = [];
+  await pipeline(
```
Promise `pipeline`s are really cool.

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+  }[] = [];
+  await pipeline(
+    startDownload(url),
+    extractTar(
```
We can use `extractTar` to only extract the binaries we asked for. So we do. Zoom zoom.

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+  });
+  const source: Source = {
+    async size() {
+      const download = startDownload(url, { agent, method: 'HEAD' });
```
We just request the headers from github, we want to know how big the zip file is so the `unzipper` package can ask for the zip's "central directory" (an index of all the files in the zip), which is always near the end of the zip file.

The `download` variable here is a `PassThrough` stream, which, for us, is just a stream that acts a middleman between the source stream and a destination stream.

I'll explain a bit more in the `startDownload` function below.

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+      const options = {
+        agent,
+        headers: {
+          range: `bytes=${offset}-${bytes ? offset + bytes : ''}`,
```
Grab only certain parts of the file from github; no need to download the whole thing!

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+  };
+
+  const { files } = await Open.custom(source, {});
+  const filtered = files.filter(({ path }) =>
+    binaries.includes(basename(path, extname(path))),
+  );
```
Gets the _index_ to the binaries we care about (no files have been downloaded yet)

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+    binaries.includes(basename(path, extname(path))),
+  );
+  return await Promise.all(
+    filtered.map(async ({ path, stream }) => {
+      const dest = join(dir, path);
+      const entry = stream();
```
Now we can download only the binaries we care about.

davidmurdoch commented on test/helpers/foundry/download.ts:
```diff
@@ -0,0 +1,82 @@
+      response.destroy();
+    } else {
+      // resolve with response stream
+      stream.emit('response', response);
+
+      response.once('error', stream.emit.bind(stream, 'error'));
+      await pipeline(response, stream).catch(stream.emit.bind(stream, 'error'));
```
More cool `pipeline` usage!

the call to `startDownload` created the `stream` earlier (a `PassThrough` steam), and returned it immediately even thought we hadn't actually downloaded anything yet. Now that we have started receiving data we just send it along.

Notice that we also emit errors on the `steam` object when appropriate.

davidmurdoch commented on test/helpers/foundry/download.ts:
```diff
@@ -0,0 +1,82 @@
+import { DownloadOptions } from './types';
+
+class DownloadStream extends Stream.PassThrough {
+  async response(): Promise<IncomingMessage> {
+    return new Promise((resolve, reject) => {
+      this.once('response', resolve);
+      this.once('error', reject);
+    });
+  }
```
In a `HEAD` only request, there is no `body`, so there is no data to send to the stream. This `response` function is used to return the headers (and `status`, `statusCode`, etc). outside of the stream itself

davidmurdoch commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+  dir: string,
+  checksumAlgorithm?: string,
+) {
+  const agent = new (url.protocol === 'http:' ? HttpAgent : HttpsAgent)({
+    keepAlive: true,
+  });
```
We end up querying the same connection multiple times, so keeping the connection alive in between queries improves performance.

davidmurdoch committed:
move tat and minipass in depcheck to the CI/testing section

davidmurdoch commented on .depcheckrc.yml:
```diff
@@ -82,6 +82,8 @@ ignores:
   - 'mini-css-extract-plugin'
   - 'webpack-cli'
+  # foundryup
+  - 'tar'
```
done

davidmurdoch commented on jest.config.js:
```diff
@@ -5,6 +5,7 @@ module.exports = {
     '<rootDir>/ui/**/*.(js|ts|tsx)',
     '<rootDir>/development/build/transforms/**/*.js',
     '<rootDir>/test/unit-global/**/*.test.(js|ts|tsx)',
+    '<rootDir>/test/helpers/foundry/**/*.(js|ts|tsx)',
```
add to test coverage

davidmurdoch commented on jest.config.js:
```diff
@@ -36,6 +37,7 @@ module.exports = {
     '<rootDir>/test/unit-global/**/*.test.(js|ts|tsx)',
     '<rootDir>/test/e2e/helpers.test.js',
     '<rootDir>/test/e2e/helpers/**/*.test.(js|ts|tsx)',
+    '<rootDir>/test/helpers/foundry/**/*.test.(js|ts|tsx)',
```
add foundryup tests to jest tests

davidmurdoch commented:
@MetaMask/security-liaisons the build-system lavamoat policy changes are just from adding a new dev dependency, `unzipper`, that gets picked up in tests.

naugtur commented:
> @MetaMask/security-liaisons the build-system lavamoat policy changes are just from adding a new dev dependency, `unzipper`, that gets picked up in tests.

the team name is @MetaMask/policy-reviewers (more future proof and easier to avid typos)

The policy change is indeed trivial, but it indicates one interestin thing - there's no policy for the package in question, so it's likely that not all code that uses our dependencies is running under policy enforcement. 

1. Does it make sense for the anvil related stuff to run under lavamoat?
1.1. If not, why?
2. Are we aware of (or have documented) which parts of our CI automation run with and without lavamoat?

// 2 is not related to this PR, just worth bringing up in context.

HowardBraham commented on test/helpers/foundry/extract.ts:
```diff
@@ -0,0 +1,211 @@
+    for (const { path, binary, checksum } of downloads) {
+      if (checksums) {
+        say(`verifying checksum for ${binary}`);
+        const expected = checksums.binaries[binary as Binary];
```
This fails on Windows because `binary = 'anvil.exe'` and the data is in `checksums.binaries['anvil']`, not `checksums.binaries['anvil.exe']`

davidmurdoch replied:
![image](https://github.com/user-attachments/assets/dc61fd3f-24bc-4434-9f54-b87c3ce8650b)

HowardBraham commented on package.json:
```diff
@@ -44,6 +45,7 @@
     "forwarder": "node ./development/static-server.js ./node_modules/@metamask/forwarder/dist/ --port 9010",
     "dapp-forwarder": "concurrently -k -n forwarder,dapp -p '[{time}][{name}]' 'yarn forwarder' 'yarn dapp'",
     "test:unit": "jest",
+    "anvil": "npx anvil",
```
This doesn't work on Windows, but `npx anvil.exe` does.  I'm not sure what the best solution for this is, I tried to research it but didn't find anything great.

davidmurdoch replied:
Fixed by using `node_modules/.bin/anvil` directly. Looks like Windows assumes the extension.

davidmurdoch replied:
![image](https://github.com/user-attachments/assets/c44be3f9-9347-4011-9730-fa45c6b3ac62)

davidmurdoch committed:
fix checksum checks

davidmurdoch committed:
reference binary directly

davidmurdoch committed:
update to stable release

davidmurdoch commented:
> > @MetaMask/security-liaisons the build-system lavamoat policy changes are just from adding a new dev dependency, `unzipper`, that gets picked up in tests.
> 
> the team name is @MetaMask/policy-reviewers (more future proof and easier to avid typos)
> 

👍 

> The policy change is indeed trivial, but it indicates one interestin thing - there's no policy for the package in question, so it's likely that not all code that uses our dependencies is running under policy enforcement.
> 
> 1. Does it make sense for the anvil related stuff to run under lavamoat?
>    1.1. If not, why?
> 2. Are we aware of (or have documented) which parts of our CI automation run with and without lavamoat?
> 
> // 2 is not related to this PR, just worth bringing up in context.

We have exactly _zero_ development/helper scripts that run under lavamoat (unless you consider the _production_ build system a helper script but even if you do... we don't run our development/testing build systems under lavamoat).

HowardBraham committed:
Merge branch 'main' into foundryup

seaona reviewed (dismissed):
incredible work, David :fire:

davidmurdoch committed:
lint fix

itsyoboieltr reviewed (approved):
LGTM!

MajorLift commented on lavamoat/build-system/policy.json:
It appears that `browserify` is being swapped out with `unzipper` without any new authorizations being added.

MajorLift reviewed (approved):
Policy change LGTM

davidmurdoch commented:
Looks like this still needs a review from "supply chain" (cc @naugtur ) and from Extension Security Team (cc @Gudahtt ).

naugtur reviewed (approved):
I don't know why the review is marked as mandatory but I reviewed and didn't see anything problematic. I am not sure why we're doing it this way in particular, but that's a conversation I'd rather have elsewhere.

desi reviewed (approved):
Based on reading through the code, all comments and seeing who has reviewed I feel ok to give this an approval on behalf groups from the codeowner rules.
